src/annoylib.h

```diff
@@ -324,7 +347,129 @@ struct Angular {
   }
 };
 
-struct Hamming {
+
+struct DotProduct : Angular {
+  template<typename S, typename T>
+  struct ANNOY_NODE_ATTRIBUTE Node {
+    /*
+     * This is an extension of the Angular node with an extra attribute for the scaled norm.
+     */
+    S n_descendants;
+    union {
+      S children[2]; // Will possibly store more than 2
+      T norm;
+    };
+    T dot_factor;
+    T v[1]; // We let this one overflow intentionally. Need to allocate at least 1 to make GCC happy
+  };
+
+  static const char* name() {
+    return "dot";
+  }
+
+  static inline size_t default_search_k(const size_t n, const size_t num_roots) {
+    return Angular::default_search_k(n, num_roots) * 10;
+  }
+
+  template<typename S, typename T>
+  static inline T distance(const Node<S, T>* x, const Node<S, T>* y, int f) {
+    return -dot(x->v, y->v, f);
+  }
```

Comment:
shouldn't this just be the euclidean distance between the vectors including the extra element?

Reply:
ok i'm very confused now, let me take a deeper look later today. i think we use the terminology "distance" for multiple separate things.

Reply:
(apologies for randomly popping up here and writing an essay)

@erikbern, I don't understand the rational behind using the euclidean distance here?

---
To try to clarify this from an API standpoint:

* The distance function is exposed to the users as follows:
```py
t = AnnoyIndex(2, 'dot')
t.add_item(0, [0, 1])
t.add_item(1, [1, 1])
d = t.get_distance(0, 1)
```
where `get_distance` is defined as follows:
```cpp
T get_distance(S i, S j) {
    return D::normalized_distance(D::distance(_get(i), _get(j), _f));
  }
```

*  The doc does not specify anything special about the distance: "Returns the distance between items `i` and `j`."

* Therefore, I believe the user would expect that `t.get_distance(v_1, v_2)` would use the distance specified in `AnnoyIndex(f, {DIST})`.
So in this case it would be `dot(vectors[v_1], vectors[v_2])`. 

---

One issue is that `dot(x, y)` is more a similarity than a distance. I assume `-dot(...)` is more convenient than `C - dot(...)` since the dot is not normalized.

I guess in this case we could return either `-dot(original_vec_i, original_vec_j)` or `1 - cos(augmented_vec_i, augmented_vec_j)`.
(where `original_vec_i` is of dimension `N` and `augmented_vec_i` is of dimension `N+1`).

Given the API example above, returning the `dot` value seems to make more sense to me than returning the `cos` value.

---

=> @erikbern, @psobot: WDYT?

Reply:
@yonromai the reason i originally brought up euclidean distance was that i thought the distance method was used internally somewhere for the splits.

if it's only for external consumption then i don't have any strong feelings, i guess positive dot product is fine